WIP LayoutEditor-More LayoutEditorConnectivityTests

geowar1 opened the pull request:
Ah! Have to change the base branch to "development"!

bobjacobsen commented:
Restarted Travis separate tests:
```
There was 1 failure:
8701) testTimedSensorFollowing(jmri.implementation.LightControlTest)
871java.lang.NullPointerException
872	at jmri.implementation.LightControl.deactivateLightControl(LightControl.java:868)
873	at jmri.implementation.AbstractLight.lambda$deactivateLight$1(AbstractLight.java:493)
874	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
875	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
876	at jmri.implementation.AbstractLight.deactivateLight(AbstractLight.java:492)
877	at jmri.implementation.LightControlTest.testTimedSensorFollowing(LightControlTest.java:620)
878	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

bobjacobsen commented:
AppVeyor timed out & restarted.  Log ended with
unit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1941    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1942    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1943    [junit] AL lib: (EE) MMDevApiOpenPlayback: Device init failed: 0x80070490
1944    [junit] java.lang.NullPointerException
1945    [junit] 	at sun.awt.windows.WWindowPeer.getActiveWindowHandles(WWindowPeer.java:463)
1946    [junit] 	at sun.awt.windows.WToolkit.eventLoop(Native Method)
1947    [junit] 	at sun.awt.windows.WToolkit.run(WToolkit.java:315)
1948    [junit] 	at java.lang.Thread.run(Thread.java:748)

geowar1 commented:
I merged in development since it has some fixes that I think improve Appveyor runtimes.

geowar1 commented:
AppVeyor: testStopButton failed (timeout).

geowar1 commented:
[junit] ERROR - testStopButton(jmri.jmrit.throttle.ThrottleFrameTest) : giving up after 4 failures [main] util.junit.rules.RetryRule.evaluate()
1834

bobjacobsen commented:
This is now showing conflicts in `java/test/jmri/jmrit/display/layoutEditor/LayoutEditorToolsTest.java` that are too complicated for me to resolve by default.

bobjacobsen commented:
AppVeyor failed without end report. Close/Reopen to pick up @pabender's fix of that.

bobjacobsen commented:
Is this ready to merge?  I'd like to get it merged as soon as CI is passed. If not, could you mark it WIP please?

geowar1 commented:
Ready to merge… just waiting for appveyor to pass.

bobjacobsen commented:
close and reopen to pick up #7761

bobjacobsen commented:
AppVeyor ended with zero failures, three errors.  Log is at https://ci.appveyor.com/project/JMRI/jmri/builds/29731015

Not restarted, pending a fix for the problem.

geowar1 commented:
[WARNING] Rule violated for source file jmri/jmrit/display/layoutEditor/LayoutEditorVerticalToolBarPanel.java: lines covered ratio is 0.00000, but expected minimum is 0.00001

Fixing via PR #7770

bobjacobsen commented:
Close and reopen to pick up #7770

geowar1 commented:
Anyone have a clue what the AppVeyor/pr failure is?

geowar1 commented:
jmri.jmrit.operations.setup.OperationsSetupFrameTest          testSetupFrameWrite <error message="Wait AbstractButton with text "HO" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.TimeoutEx...
1806jmri.jmrit.operations.automation.actions.BuildTrainActionTest testActionMessages  <error message="Wait AbstractButton with text "Halt" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.Timeout...
1807jmri.jmrit.timetable.swing.TimeTableFrameTest                 testDriver          <error message="Wait "Item "Station 2" available in combo box." state to be reached (ComponentOperator.WaitStateTimeout)" ty...

geowar1 commented:
jmri.jmrit.operations.setup.OperationsSetupFrameTest          testSetupFrameWrite <error message="Wait AbstractButton with text "HO" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.TimeoutEx...
1808jmri.jmrit.operations.automation.actions.BuildTrainActionTest testActionMessages  <error message="Wait AbstractButton with text "Halt" loaded (ComponentOperator.WaitComponentTimeout)" type="org.netbeans.jemmy.Timeout...
1809jmri.jmrit.timetable.swing.TimeTableFrameTest                 testDriver          <error message="Wait "Item "Station 2" available in combo box." state to be reached (ComponentOperator.WaitStateTimeout)" ty...

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorAuxToolsTest.java:
```diff
@@ -517,6 +519,8 @@ public static void setUpClass() throws Exception {
             JUnitUtil.initInternalSensorManager();
 
+            //save the old string comparator
+            stringComparator = Operator.getDefaultStringComparator();
```
I don’t see where you are restoring this in this file.

geowar1 replied:
Thanks for the catch… (now fixed) I'm actually trying to figure things out in PR #7264. Eyeballs there would be welcome.

pabender commented on java/test/jmri/jmrit/display/layoutEditor/LayoutEditorTest.java:
```diff
@@ -919,7 +919,7 @@ public void testToolBarPositionLeft() {
                 + Bundle.getMessage("ToolBarSideLeft"), "/");
 
-        delay(500); //TODO: switch to JemmyUtil.delay
+        JemmyUtil.delay(500); //TODO: switch to JemmyUtil.delay
```
Make sure you remove the todo comments from these lines.

Also, I am a bit concerned that you have a wrapper around a sleep here.  It should be sufficient for any graphical operation to use the Jemmy QueueTool to wait for the awt event queue to be empty.  Raw half second delays in tests are a code smell.

geowar1 replied:
That's great in theory… in practice however I'm finding that sometimes neither QueueTool().waitEmpty() nor EventTool().waitNoEvent() work… the 1/2 second was somewhat arbitrary. "Code smell" is an excellent description of Jemmy. I'm finding it woefully under documented. For example: what's the difference between .push() and .doClick()? In JennyUtils three of the four enterClickAndLeave methods used .doClick() and the other .push(). Why? Changing them all to use .push broke the tests so I changed them all to .doClick() (which only makes sense given the name of the methods). But when to use one vs. the other? Not a clue.

pabender replied:
doClick and push cause different events to be sent to the event queue.  Some events are triggered by a mouse down event ( which both send ) some events are triggered by a mouse up event ( doClick sends that ).  doClick is roughly the equivalent of a push followed by a release.

Where you really have to be careful with this is that some actions don’t respond correctly if the button has already been released when the event is handled ( I.e. if too little time occurs between the mouse up and mouse down ).  The use of push in the JemmyUtils was intentional where it is used.

geowar1 replied:
There's a setting for the time between the press and the release… I suspect it's a global setting so if it's changed it should be changed back so that it won't effect other tests.

geowar1 replied:
All delay/sleep code has been REMOVED.

pabender commented on java/test/jmri/util/swing/JemmyUtil.java:
```diff
@@ -1,120 +1,471 @@
+        for (long delta = milliseconds; delta > 0;
+                delta = nextTime - System.currentTimeMillis()) {
+            try {
+                Thread.sleep(delta);
```
As I said in a previous comment, we really need to rethink this.  I don’t like the idea of having a raw sleep call in any thread during a test.  It only causes the test to take longer.

geowar1 replied:
Case in point (sometimes QueueTool().waitEmpty() nor EventTool().waitNoEvent() don't work): In LayoutEditorTests the testToolBarPosition[Left, Bottom, Right, Float] methods all fail AppVeyor if I use the wait methods instead of the delay. (Lower values may work… but I have't tried them. 1/2 second was a WAG.)

pabender replied:
I am not going to approve this with a raw sleep in place.

geowar1 replied:
Good for you. Do you want me to give up on trying to make LE maintainable and improving code coverage? I've spent the better part of December trying to do both with virtually non-existent Jemmy documentation and dealing with all these waitXXX timeouts that should "just work". I agree that NONE of these wait(/delay) methods should be necessary… but they are. Would you feel better if the code limited the sleep delay to 1/4 second? Or smaller?

geowar1 replied:
The place where I found sleep to be necessary was the LayoutEditorTest:testToolBarPosition[Left, Bottom, Right, Float] methods. Feel free to give them a look and suggest what they should be waiting for instead.

pabender replied:
Thinking about this some more, I realized what the more critical problem I have with this is.  There is already a waitFor method in JUnitUtil that waits for a time period in microsecods.

So remove the delay method because it is duplicative, replace uses of it with JUnitUtil.waitFor(int).

We would still prefer to use one of the waitFor methods with a condition if it is at all possible.

Also, the Jemmy operators do wait for a period of time for  an object to appear.  Delays for GUI operations checked with Jemmy operators really should all be handled by Jemmy itself.

Jemmy has a method to get and set timeouts specified by the Timeoutable interface.  https://javadoc.io/doc/org.netbeans/jemmy/latest/org/netbeans/jemmy/Timeoutable.html

Most, if not all, of the operators implement this interface.  You should probably use that instead of an external delay.

geowar1 replied:
> Also, the Jemmy operators do wait for a period of time for an object to appear. Delays for GUI operations checked with Jemmy operators really should all be handled by Jemmy itself.

That 110% great… in theory… In practice the testToolBarPositionXXX tests were failing without the (short) 500 mSec delay before the next (waitFor) operation… that would then timeout after 10 seconds.

At this point I'm content to remove all the delay/sleep code and mark those methods @Ignore and move on… I'll "resolve" this after my next commit.

geowar1 replied:
Done! All testToolBarPositionXXX tests now use "new EventTool().waitNoEvent(200);" instead of delay and all delay/sleep code has been removed from JemmyUtil.

bobjacobsen commented:
> Case in point (sometimes QueueTool().waitEmpty() nor EventTool().waitNoEvent() don't work): In LayoutEditorTests the testToolBarPosition[Left, Bottom, Right, Float] methods all fail AppVeyor if I use the wait methods instead of the delay. (Lower values may work… but I have't tried them. 1/2 second was a WAG.)

What are you trying to test?

If you’re testing the _function_ behind some control, then just don’t do that through the GUI.  Test it directly instead. If that’s hard, then the code should be restructured to proper MVC so it’s easier.

If you’re testing that the GUI is connected to the appropriate code, then you can try the jmri.util.JUnitUtil.waitFor(..) method and inheritance to mock out whatever the GUI is calling.

geowar1 commented:
MVC is great in theory… but LE isn't remotely in the neighborhood. And I really don't want to start refactoring LE until I've got code coverage much closer to 100%. The test is calling setToolBarSide with left, bottom, right & float (always returning to top before the next test. The waitFor(), QueueTool waitEmpty nor EventTool waitNoEvent methods were able to get a reliable test. The 500 mSec sleep works.

pabender commented:
> MVC is great in theory… but LE isn't remotely in the neighborhood. And I really don't want to start refactoring LE until I've got code coverage much closer to 100%. The test is calling setToolBarSide with left, bottom, right & float (always returning to top before the next test. The waitFor(), QueueTool waitEmpty nor EventTool waitNoEvent methods were able to get a reliable test. The 500 mSec sleep works.

And that is why this delay concerns me.  We have methods available to us that should measure the progress and/or completion of the operation ( waitFor and waitEmpty ) yet you are using an arbitrary delay to hope the condition is satisfied.

So, are you measuring the right condition?  What signifies the operation was successful?

Your delay SHOULD be followed by an assertion.  The assertion condition can be your condition in the waitFor call.

bobjacobsen commented:
Complex Swing operations, like laying out a window with a new shape, can take _many_ sequential events: things change, which fire events, which cause changes, which fire more events, etc. Depending on the L&F, these might not be consecutive: work can go to the native system, take some time on another thread, then come back and trigger more Swing events. That might even vary by platform. 

So I’ll make my point again: if these tests are just exercising GUI layout, without doing any actual checks of that _layout_, why automate them?  Why do you need to check LayoutEditor _function_ with these different orientations?

geowar1 commented:
> Why do you need to check LayoutEditor function with these different orientations?

Code coverage

pabender commented:
> > Why do you need to check LayoutEditor function with these different orientations?
> 
> Code coverage

Code coverage is only a PART of the testing picture.  You also have to be able to check the post-conditions of any test case.  Any good test starts with a set of pre-condition, performs some operation ( possibly with other inputs ) and produces a set of post-conditions.

Without actually checking something meaningful, the test is really an invalid test, which brings me back to my earlier questions:

What signifies the operation was successful?
Is the current test measuring the right post-condition?

geowar1 commented:
> Code coverage is only a PART of the testing picture.

I 100% get this… OTOH LayoutEditor.java's code coverage is 46%; That's an "F".

How do I improve this when most of the code is ether setup or drawing code that doesn't have a realistically testable end state?
(Same thing with panel.js).

IOW: I may not have a way of testing the results of the setToolBarPosition method… but I can at least improve the code coverage above an "F".

Plus: You're complaining about 4(!) tests out of 103 because they're not perfect? When you start at "unmaintainable" you don't aim for "perfect"… you try for "better".

Seriously, would you rather I not work on testing? I do have better things to do. You're perfectly welcome to write all the test code yourself.

If you really actually would like to be helpful tell me how to run the AppVeyor test locally so I don't have to wait (an hour+) for it to fail on GitHub.

I tried "term> mvn logall" on the "development" branch… it took 13:41:27.03s hours and had 142 errors and 29 failures out of 54,708 tests.

Here's the top 7 worse timeouts:
 1,080.928 s <<< FAILURE! - in jmri.jmrit.display.PositionableLabelTest
 1,095.533 s <<< FAILURE! - in jmri.jmrit.consisttool.ConsistToolFrameTest
 1,280.592 s <<< FAILURE! - in jmri.jmrit.display.layoutEditor.MemoryIconTest
 1,441.226 s <<< FAILURE! - in jmri.jmrit.display.BlockContentsIconTest
 1,452.066 s <<< FAILURE! - in jmri.jmrit.display.MemoryIconTest
 2,233.897 s <<< FAILURE! - in jmri.jmrit.logix.LinkedWarrantTest
26,963.078 s <<< FAILURE! - in jmri.jmrit.operations.automation.AutomationCopyFrameTest
(Why are these so high? isn't 180s the default timeout value? None of these tests change the global timeout.)

Then I tried this:

term> mvn clean test exec:java -Dexec.classpathScope="test" -Dexec.mainClass="org.hid4java.UsbHidDeviceExample"

(I don't remember where I got this command…)

On the "development" branch this took a little over an hour. (I didn't save the log so I don't have the errors/failures… but it wasn't zero/zero (which doesn't make me feel good).)

bobjacobsen commented:
> I 100% get this… OTOH LayoutEditor.java's code coverage is 46%; That's an "F".
> 
> How do I improve this when most of the code is ether setup or drawing code that doesn't have a realistically testable end state?
> (Same thing with panel.js).

It’s great that you want to get this code more reliable.  I _definitely_ agree that it’s better to have tests in place before changing the code.

I don’t think that the fraction shown by CodeCoverage is a good metric.  46% of functional code covered wouldn’t be great, as then changes could break things.  On the other hand, a _lot_ of Swing code is just boilerplate callbacks, layout, etc. Getting high true coverage (not just running, but checking the results) is close to impossible.  

So, at the risk of getting another curt reply, let me recommend that you just @Ignore these complex GUI checks, and move to checks that focus on the operation of the part of the code that you’d like to improve. In doing that, you might have to make some changes to make things more testable (smaller routines, clearer context), but that’s a win-win. 

Again, it’s great that you want to make LayoutEditor code more understandable and reliable.  I’m sorry this has been so frustrating.

geowar1 commented:
The job exceeded the maximum time limit for jobs, and has been terminated.
(Both Travis & AppVeyor timed out after an hour…)

testSaveCarHazardous(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testSaveCarUtility(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testWeightErrorConditions(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testLocationComboBox(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testSaveExistingCar(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testSaveCarCaboose(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testSaveCarPassenger(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testSaveKernel(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testRoadNumberErrorConditions(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()
testAddCar(jmri.jmrit.operations.rollingstock.cars.CarEditFrameTest) : giving up after 3 failures [main] util.junit.rules.RetryRule.evaluate()

geowar1 commented:
Many of the LayoutEditor sources have the worse code coverage scores in JMRI. I'd like to at the very least get all LE source better than an "F"… preferable "A" (yeah, I'm dreaming…)
Note: IIRC LayoutEditor.java's code coverage used to be 8%. It has taken a LOT of work to get it up to 46%. Probably got most of it by moving all the drawing code to LayoutEditorComponent and all the ToolBarPanel into LayoutEditorToolBarPanel (and subclasses). Before that I moved all the dialog code out of LE and into the LayoutEditorDialog source (now in it's own sub-package). Before that I moved all the LayoutTrack specific code into LayoutTrack and its sub-classes. All the while trying to not break anything. It's been a PITA. But hopefully for the improvement of the LE package. Oh, and a few new "features" have been slipped in also.
--
I'm going to take your suggestion to heart: I'm not going to try to fix any of the tests that are timing out now… I'm just going to mark them @Ignore and move on.
Code coverage "D" or bust! ;-)
--
(BTW: I haven't said it enough: I do appreciate you taking the time to help… it's just frustrating. I try not to bother you but the available jemmy documentation is especially poor (as I'm sure I've mentioned).)

pabender commented:
> Many of the LayoutEditor sources have the worse code coverage scores in JMRI. I'd like to at the very least get all LE source better than an "F"… preferable "A" (yeah, I'm dreaming…)
> 
> Note: IIRC LayoutEditor.java's code coverage used to be 8%. It has taken a LOT of work to get it up to 46%. Probably got most of it by moving all the drawing code to LayoutEditorComponent and all the ToolBarPanel into LayoutEditorToolBarPanel (and subclasses). Before that I moved all the dialog code out of LE and into the LayoutEditorDialog source (now in it's own sub-package). Before that I moved all the LayoutTrack specific code into LayoutTrack and its sub-classes. All the while trying to not break anything. It's been a PITA. But hopefully for the improvement of the LE package. Oh, and a few new "features" have been slipped in also.
> 

I certainly appreciate the effort to both make layout editor more understandable and more tested.  

> I'm going to take your suggestion to heart: I'm not going to try to fix any of the tests that are timing out now… I'm just going to mark them @ignore and move on.
> Code coverage "D" or bust! ;-)

Right, sometimes it is best to just ignore a test and move on.  You, or someone else, can always come back to it later.

> (BTW: I haven't said it enough: I do appreciate you taking the time to help… it's just frustrating. I try not to bother you but the available jemmy documentation is especially poor (as I'm sure I've mentioned).)

I understand the frustration.  My knowledge of how Jemmy works has been earned over the last 4 years as I have worked on generally increasing the coverage in JMRI.  At times, the best references turn out to be existing usage of Jemmy code in JMRI.
